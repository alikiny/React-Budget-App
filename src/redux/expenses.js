import { v4 as uuid } from 'uuid';
import { firebase } from '../firebase/firebase'

export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
})

export const setExpensesAction = () => {
    const expenses = []
    return (dispatch) => {
        return firebase.database().ref('expenses')
            .once('value').then(snapshot => {
                snapshot.forEach((expense) => {
                    expenses.push({
                        id: expense.key,
                        ...expense.val()
                    })
                })

                dispatch(setExpenses(expenses))
            })

    }


}

export const addExpense = (expense) => ({
    type: 'ADD_EXPENSES',
    expense
})

export const startAddAction = (expense = {}) => {

    return (dispatch) => {
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0
        } = expense

        const newExpense = { description, note, amount, createdAt }
        return firebase.database().ref('expenses').push(newExpense)
            .then((ref) => {
                dispatch(addExpense({
                    id: ref.key,
                    ...newExpense
                }))
            })
    }

}

export const removeExpense = (id) => ({
    type: 'REMOVE_EXPENSE',
    expense: {
        id
    }
})

export const editExpense = (id, update) => ({
    type: 'EDIT_EXPENSE',
    expense: {
        id,
        update
    }
})


const defaultExpenses = []
export const expensesReducer = (state = defaultExpenses, action) => {
    switch (action.type) {
        case 'SET_EXPENSES':
            return action.expenses
        case 'ADD_EXPENSES':
            return [...state, action.expense]

        case 'REMOVE_EXPENSE':
            return state.filter((expense) => expense.id !== action.expense.id)

        case 'EDIT_EXPENSE':
            return state.map((expense) => {
                if (expense.id === action.expense.id) {
                    console.log('Editted successfully!')
                    return { ...expense, ...action.expense.update }

                }
                else {

                    return expense
                }
            }
            )

        default:
            return state
    }
}