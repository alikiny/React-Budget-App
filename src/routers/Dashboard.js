import React, { useState } from 'react';
import AccountInfo from '../dashboard-component/AccountInfo'
import ProfileInfo from '../dashboard-component/ProfileInfo'
import SavingInfo from '../dashboard-component/SavingInfo'
import SettingInfo from '../dashboard-component/SettingInfo'
import { connect } from 'react-redux';



export const Dashboard = (props) => {

    const navigateComponent = (componentName) => {

    }

    const [DynamicComponent, setComponent] = useState(<ProfileInfo/>)

    const onChangeDynamicComponent = (option) => {
        switch (option) {
            case 'profile':
                setComponent(<ProfileInfo/>);
                break;
            case 'setting':
                setComponent(<SettingInfo/>);
                break;
            case 'account':
                setComponent(<AccountInfo/>);
                break;
            case 'saving':
                setComponent(<SavingInfo/>);
                break;
            default:
                setComponent(<AccountInfo/>);
        }
    }

    return (
        <div className="container-fluid mt-5">
            <div className='row'>

                <div className="col-sm-4">
                    <ul className="list-group list-group-flush">
                        <li 
                        className="navigataion-bar list-group-item"
                        onClick={()=>onChangeDynamicComponent('profile')} >Profile</li>
                        <li 
                        className="navigataion-bar list-group-item"
                        onClick={()=>onChangeDynamicComponent('setting')}>Setting</li>
                        <li 
                        className="navigataion-bar list-group-item"
                        onClick={()=>onChangeDynamicComponent('account')}>Current account</li>
                        <li className="navigataion-bar list-group-item"
                        onClick={()=>onChangeDynamicComponent('saving')}>Saving goal</li>
                    </ul>
                </div>

                <div className="col-sm-8">
                    {DynamicComponent}
                </div>

            </div>
        </div>

    )
}


export default connect(state => ({
    expenses: state.expenses
}))(Dashboard)