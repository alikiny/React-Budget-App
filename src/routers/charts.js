import React from "react";
import {
Line,
LineChart,
XAxis,
YAxis,
Label,
ResponsiveContainer,
CartesianGrid,
Tooltip,
} from "recharts";


const LChart = ({ data }) => {
  return (
    <ResponsiveContainer width="100%" height={400}>
      <LineChart
        data={data}
        margin={{ top: 30, right: 30, left: 30, bottom: 30 }}>
        <CartesianGrid />
        <Tooltip />
        <XAxis type="category" dataKey="time" domain={["dataMin", "dataMax"]}>
          <Label
            value={"Time"}
            position="bottom"
            style={{ textAnchor: "middle" }}
          />
        </XAxis>
        <YAxis>
          <Label
            value={"Expenses"}
            position="left"
            angle={-90}
            style={{ textAnchor: "middle" }}
          />
        </YAxis>
        <Line
          dataKey="amount"
          name="Spent"
          unit={"EUR"}
          dot={false}
          type={"natural"}
        />
      </LineChart>
    </ResponsiveContainer>
  );
};

export default LChart;