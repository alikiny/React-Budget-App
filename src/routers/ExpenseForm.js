import React, { useState, useEffect } from 'react';
import { makeStyles } from '@material-ui/core/styles';
import Modal from '@material-ui/core/Modal';
import TextField from '@material-ui/core/TextField';
import {
    MuiPickersUtilsProvider,
    KeyboardDatePicker
} from '@material-ui/pickers'
import DateFnsUtils from '@date-io/date-fns';
import IconButton from '@material-ui/core/IconButton';
import AddCircleIcon from '@material-ui/icons/AddCircle';
import CancelIcon from '@material-ui/icons/Cancel';
import { connect } from 'react-redux';
import { startEditExpense,startAddAction } from '../redux/expenses'


const useStyles = makeStyles((theme) => ({
    paper: {
        width: 400,
        backgroundColor: theme.palette.background.paper,
        border: '2px solid #000',
        margin: 'auto',
        boxShadow: theme.shadows[5],
        padding: theme.spacing(2, 4, 3),

        '& .MuiFormControl-root': {
            margin: '10px',
            width: '100%'
        }
    },
    addIcon: {
        color: 'white',
        backgroundColor: '#4c2341'
    }
}));

export const ExpenseModal = (props) => {
    const expense = props.expense

    const [name, setName] = useState('')
    const [amount, setAmount] = useState('')
    const [note, setNote] = useState('')
    const [date, setDate] = useState(new Date(Date.now()))

    useEffect(() => {
        if (expense) {
            setName(expense.name)
            setAmount(expense.amount)
            setNote(expense.note)
            setDate(expense.date)
        }
    })

    const classes = useStyles()
    const onSave = () => {


        const alertText = []
        const nameList = props.expenses.map((ex) => ex.description)
        const itemAmount = parseFloat(amount)
        const getDate = new Date(date)
        const getTime = getDate.getTime()


        if (Number.isNaN(itemAmount)) {
            alertText.push("Amount is required")
        }


        if (name.length < 1) {
            alertText.push("Name is required")
        } else if (!props.edit && nameList.includes(name)) {
            alertText.push("Name is already taken")
        }

        //Check if there is no error, continue to add or edit expenses

        if (alertText.length < 1) {

            //Check if the command is to edit or add expense
            if (props.edit) {

                
                const itemToDelete = props.expenses.filter((e) => {
                    return e.id === props.expense.id
                })

                const toBeDeleted=itemToDelete[0]
                
                if (
                    toBeDeleted.description === name
                    && toBeDeleted.amount ==amount
                    && toBeDeleted.note ===note
                    && toBeDeleted.createdAt ==getTime
                ) {
                    alert('There is no change made')
                } else {
                    console.log('edit modal')
                    props.startEditExpense(
                        toBeDeleted.id,
                        {
                            description: name,
                            note,
                            amount:itemAmount,
                            createdAt: getTime
                        }
                    )
                    
                }


                //Add expense incase 'edit ' is set to false
            } else {
                props.addExpense({
                    description: name,
                    note,
                    amount:itemAmount,
                    createdAt: getTime
                })
                console.log('added new expense')
            }

            props.closeModal()
            
        } 
        /* else {
            alert(alertText)
        } */
        setName('')
        setAmount('')
        setNote('')
        setDate(new Date(Date.now()))

        

    }


    return (
        <Modal
            open={props.openModal}
            onClose={props.closeModal}>
            <div className={classes.paper}>
                <TextField
                    label="Name"
                    id='item-name'
                    value={expense ? expense.name : name}
                    onChange={(e) => {
                        if (expense) {
                            expense.name = e.target.value
                        }
                        setName(e.target.value)
                    }
                    }
                    variant="outlined"
                ></TextField>

                <TextField
                    label="Amount"
                    id='item-amount'
                    type="number"
                    value={expense ? expense.amount : amount}
                    onChange={(e) => {
                        if (expense) { expense.amount = e.target.value }
                        setAmount(e.target.value)

                    }}
                    variant="outlined"
                ></TextField>

                <MuiPickersUtilsProvider utils={DateFnsUtils}>
                    <KeyboardDatePicker
                    id='item-date'
                        disableToolbar
                        inputVariant="outlined"
                        format="MM/dd/yyyy"
                        margin="normal"
                        value={expense ? expense.date : date}
                        label="Date"
                        onChange={
                            (date) => {
                                setDate(date)
                                if(expense){
                                    expense.date = date
                                }
                                
                            }
                        }
                        KeyboardButtonProps={{
                            'aria-label': 'set date',
                        }}
                    />
                </MuiPickersUtilsProvider>

                <TextField
                    label="Note"
                    id='item-note'
                    value={expense ? expense.note : note}
                    onChange={(e) => {
                        setNote(e.target.value)
                        if(expense){expense.note = e.target.value}
                    }}
                    variant="outlined"
                ></TextField>

                <div>
                    <IconButton id="save-btn" onClick={() => onSave()}>
                        <AddCircleIcon className={classes.addIcon} />
                    </IconButton>
                    <IconButton id="close-btn" onClick={() => props.closeModal()}>
                        <CancelIcon className={classes.addIcon} />
                    </IconButton>
                </div>



            </div>



        </Modal>
    )
}

const mapDispatchToProps=dispatch=>({
    addExpense: (expense)=>dispatch(startAddAction(expense)),
    startEditExpense:(id,expense)=>dispatch(startEditExpense(id,expense))
})

const mapStateToProps=state=>({
    expenses:state.expenses
})
export default connect(mapStateToProps,mapDispatchToProps)(ExpenseModal)