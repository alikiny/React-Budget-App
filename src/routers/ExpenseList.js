import React, { useState } from 'react'
import { connect } from 'react-redux'
import { getExpenses } from '../redux/selectors'
import { makeStyles } from '@material-ui/core/styles';
import Table from '@material-ui/core/Table';
import TableRow from '@material-ui/core/TableRow';
import TableBody from '@material-ui/core/TableBody';
import TableCell from '@material-ui/core/TableCell';
import TableContainer from '@material-ui/core/TableContainer';
import TablePagination from '@material-ui/core/TablePagination';
import Checkbox from '@material-ui/core/Checkbox';
import Paper from '@material-ui/core/Paper';
import { CustomToolbar } from './TooBar'
import { CustomTableHead } from './TableHead'
import { removeExpense } from '../redux/expenses'
import Tooltip from '@material-ui/core/Tooltip';
import IconButton from '@material-ui/core/IconButton';
import ArrowUpwardIcon from '@material-ui/icons/ArrowUpward';
import moment from 'moment'
import ExpenseModal from './ExpenseForm'


function createExpense(id, name, amount, date, note) {
    return { id, name, amount, date, note }

}


function EnhancedTable(props) {
    const rows = props.rows
    const classes = useStyles();
    const [selected, setSelected] = React.useState([]);
    const [page, setPage] = React.useState(0);
    const [dense, setDense] = React.useState(false);
    const [rowsPerPage, setRowsPerPage] = React.useState(5);
    const [openModal, setOpen] = useState(false)
    const [currentExpense, setExpense] = useState()
    const [isEdit, setEdit] = useState(true)
    const closeModal = () => {
        setOpen(false)
    }

    const handleSelectAllClick = (event) => {
        if (event.target.checked) {
            const newSelecteds = rows.map((n) => n.name);
            setSelected(newSelecteds);
            return;
        }
        setSelected([]);
    };

    const handleClick = (event, name) => {
        const selectedIndex = selected.indexOf(name);
        let newSelected = [];

        if (selectedIndex === -1) {
            newSelected = newSelected.concat(selected, name);
        } else if (selectedIndex === 0) {
            newSelected = newSelected.concat(selected.slice(1));
        } else if (selectedIndex === selected.length - 1) {
            newSelected = newSelected.concat(selected.slice(0, -1));
        } else if (selectedIndex > 0) {
            newSelected = newSelected.concat(
                selected.slice(0, selectedIndex),
                selected.slice(selectedIndex + 1),
            );
        }

        setSelected(newSelected);
    };

    const handleChangePage = (event, newPage) => {
        setPage(newPage);
    };

    const handleChangeRowsPerPage = (event) => {
        setRowsPerPage(parseInt(event.target.value, 10));
        setPage(0);
    };

    const handleChangeDense = (event) => {
        setDense(event.target.checked);
    };

    const isSelected = (name) => selected.indexOf(name) !== -1;

    const match = (array1, array2) => {
        const tobeDelete = []
        array1.forEach(item1 => {
            array2.forEach((item2) => {
                if (item2.name === item1) {
                    tobeDelete.push(item2)

                }
            })
        })
        return tobeDelete
    }

    const onDelete = () => {
        const tobeDelete = match(selected, rows)
        tobeDelete.forEach(item =>
            props.deleteItem(item)
        )
        setSelected([])
    }

    const emptyRows = rowsPerPage - Math.min(rowsPerPage, rows.length - page * rowsPerPage);

    return (
        <div className={classes.root}>
            <ExpenseModal openModal={openModal} closeModal={closeModal} expense={currentExpense} edit={isEdit} />
            <Paper className={classes.paper}>
                <CustomToolbar numSelected={selected.length} classes={classes} onDelete={onDelete} />
                <TableContainer>
                    <Table
                        className={classes.table}
                        aria-labelledby="tableTitle"
                        size={dense ? 'small' : 'medium'}
                        aria-label="enhanced table"
                    >
                        <CustomTableHead
                            classes={classes}
                            numSelected={selected.length}
                            onSelectAllClick={handleSelectAllClick}
                            rowCount={rows.length}
                        />

                        <TableBody>
                            {rows.length === 0 && (
                                <tr>
                                    <td colspan="6">There is no expense</td>
                                </tr>
                            )}
                            {rows.map((row, index) => {
                                const isItemSelected = isSelected(row.name);
                                const labelId = `enhanced-table-checkbox-${index}`;
                                let date = new Date(row.date)
                                const newdate = moment(date).format("dddd, MMMM Do YYYY, h:mm:ss a")

                                return (

                                    <TableRow
                                        hover

                                        role="checkbox"
                                        aria-checked={isItemSelected}
                                        tabIndex={-1}
                                        key={index}
                                        selected={isItemSelected}
                                    >

                                        <TableCell padding="checkbox">
                                            <Checkbox
                                                onClick={(event) => handleClick(event, row.name)}
                                                checked={isItemSelected}
                                                inputProps={{ 'aria-labelledby': labelId }}
                                            />
                                        </TableCell>
                                        <TableCell component="th" id={labelId} scope="row" padding="none">
                                            {row.name}
                                        </TableCell>
                                        <TableCell align="left">{row.amount}</TableCell>
                                        <TableCell align="left">{newdate}</TableCell>
                                        <TableCell align="left">{row.note}</TableCell>
                                        <TableCell align="left" onClick={
                                            () => {
                                                setExpense(row)
                                                setOpen(true)
                                            }
                                        }>Edit</TableCell>
                                    </TableRow>
                                );
                            })}
                            {emptyRows > 0 && (


                                <TableRow style={{ height: (dense ? 33 : 53) * emptyRows }}>
                                    <TableCell colSpan={6} />
                                </TableRow>



                            )}
                        </TableBody>
                    </Table>
                </TableContainer>
                <TablePagination
                    rowsPerPageOptions={[5, 10, 25]}
                    component="div"
                    count={rows.length}
                    rowsPerPage={rowsPerPage}
                    page={page}
                    onChangePage={handleChangePage}
                    onChangeRowsPerPage={handleChangeRowsPerPage}
                />
            </Paper>



            <Tooltip title='Back to top' style={styles.backToTop}>
                <IconButton onClick={() => window.scroll(0, 0)}>
                    <ArrowUpwardIcon />
                </IconButton>
            </Tooltip>


        </div>
    );
}

export const ExpenseList = (props) => {

    const deleteItem = (expense) => {
        props.dispatch(removeExpense(expense.id))
    }

    const rows = props.expenses.map((expense, index) => {
        return createExpense(expense.id, expense.description, expense.amount,
            expense.createdAt, expense.note)
    })

    return (
        <div>
            <div style={styles.expenseList}>
                <EnhancedTable rows={rows} deleteItem={deleteItem} />
            </div>

        </div>


    )
}

export default connect((state) => {
    return {
        expenses: getExpenses(state.expenses, state.filters),
    }
})(ExpenseList)

const styles = {
    expenseList: {
        color: 'white',
        textAlign: 'center'
    },
    backToTop: {
        position: 'fixed',
        bottom: '10%',
        right: '2%'
    }
}

const useStyles = makeStyles((theme) => ({
    root: {
        width: '100%',
    },
    paper: {
        width: '100%',
        marginBottom: theme.spacing(2),
    },
    table: {
        minWidth: 750,
    },
}));