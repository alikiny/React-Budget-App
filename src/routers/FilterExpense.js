
import React, { useState } from 'react';
import TextField from '@material-ui/core/TextField';
import { makeStyles } from '@material-ui/core/styles';
import { connect } from 'react-redux';
import { MenuItem } from '@material-ui/core';
import Button from '@material-ui/core/Button';
import {
    MuiPickersUtilsProvider,
    KeyboardDatePicker
} from '@material-ui/pickers'
import DateFnsUtils from '@date-io/date-fns';

import { datetFilter, sortFilter, textFilter } from '../redux/filters'
import {getExpenses} from '../redux/selectors'

const useStyles = makeStyles((theme) => ({
    root: {
        background:'#25001b',
        textAlign:'center',
        padding:'20px',
        margin:'20px',
        '& .MuiTextField-root': {
            margin: theme.spacing(3),
            width: '25ch',
            backgroundColor:'white'
            
        },
        '& .MuiFormLabel-root':{
            color:'black'
        },
        '& .MuiInputLabel-outlined.MuiInputLabel-shrink':{
            transform: 'translate(20px, -20px) scale(0.75)',
            color:'white'
        }
    },
    form:{
        backgroundColor:'rgba(255,255,255,0.2)',
    },
    searchButton:{
        backgroundColor: '#4c2341',
        color:'white',
        marginTop:theme.spacing(6)
    }
}));



export const FilterExpense = (props) => {
   const[sort,setSort]=useState('Date')
    const classes = useStyles()
    const onSearch = () => {
    
        props.dispatch(textFilter(props.filters.text))
        props.dispatch(sortFilter(props.filters.sortBy))
        props.dispatch(datetFilter(props.filters.startDate,props.filters.endDate))
        console.log(getExpenses(props.expenses,props.filters))
    }

    const handleStartDateChange = (date) => {
        props.filters.startDate=date
    }

    const handleEndDateChange = (date) => {
        props.filters.endDate=date
    }

    const handleChange = (e) => {
      
        props.filters.sortBy=e.target.value   
        setSort(props.filters.sortBy)
    }

    const handleTextChange = (e) => {
        props.filters.text=e.target.value
       
    }
    return (
        <div className={classes.root}>
            <form className={classes.form}>

                <TextField
                    label="Text filter"
                    defaultValue={props.filters.text}
                    onChange={(e)=>handleTextChange(e)}
                    id="outlined-start-adornment"
                    variant="outlined"
                ></TextField>


                <TextField
                    label="Sort By"
                    select
                    value={sort}
                    onChange={handleChange}
                    id="outlined-start-adornment"
                    variant="outlined"
                >
                    <MenuItem value='Date'>Date</MenuItem>
                    <MenuItem value='Amount'>Amount</MenuItem>
                </TextField>

                <MuiPickersUtilsProvider utils={DateFnsUtils}>

                    <KeyboardDatePicker
                        disableToolbar
                        inputVariant="outlined"
                        format="MM/dd/yyyy"
                        margin="normal"
                        
                        label="Start Date"
                        value={props.filters.startDate}
                        onChange={handleStartDateChange}
                        KeyboardButtonProps={{
                            'aria-label': 'change start date',
                        }}
                    />

                    <KeyboardDatePicker
                        disableToolbar
                        inputVariant="outlined"
                        format="MM/dd/yyyy"
                        margin="normal"
                       
                        label="End Date"
                        value={props.filters.endDate}
                        onChange={handleEndDateChange}
                        KeyboardButtonProps={{
                            'aria-label': 'change end date',
                        }}
                    />
                </MuiPickersUtilsProvider>

            </form>
            <Button variant="contained" className={classes.searchButton} onClick={onSearch}>
                Search
    </Button>
        </div>
    )
}

export default connect((state) => {
    return {
        filters: state.filters,
        expenses:state.expenses
    }
})(FilterExpense)