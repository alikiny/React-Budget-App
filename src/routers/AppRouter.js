import React from 'react';
import {Route, Switch,Router } from 'react-router-dom';
import { createBrowserHistory } from 'history';
import PrivateRoute from './PrivateRoute'

import Expenses from './Expense'
import Dashboard from './Dashboard'
import Summary from './Summary'
import NotFound from './404'
import Header from './Header'


/* make sure history version compatible with react-router-dom version. Here useing 
history v4 and react-router-dom v5 */
export const history=createBrowserHistory()





const AppRouter = () => {

    
    return(
        <div>
        <Router history={history}>
            <div>
                <Header />
                <Switch>
                    <PrivateRoute path='/' component={Dashboard} exact={true} />
                    <PrivateRoute path="/expenses" component={Expenses} exact={true} />
                    <PrivateRoute path="/dashboard" component={Dashboard} exact={true} />
                    <PrivateRoute path="/summary" component={Summary} exact={true} />
                    <PrivateRoute component={NotFound} />
                </Switch>
            </div>

        </Router>
    </div>
)}
export default AppRouter;