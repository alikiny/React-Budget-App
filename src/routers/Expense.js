import React from 'react';
import ExpenseList from './ExpenseList'
import FilterExpense from '../routers/FilterExpense'



const Expenses = () => (
    <div>
        <FilterExpense />
        <ExpenseList />
    </div>

)


export default Expenses
