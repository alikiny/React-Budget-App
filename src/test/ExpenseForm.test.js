import ShallowRenderer from 'react-test-renderer/shallow'
import React from 'react'

import { ExpenseModal } from '../routers/ExpenseForm'
import expenses from './expensesTest'
import { shallow, mount } from 'enzyme'

beforeEach(() => {
    //Mock the new Date(Date.now()) function, which is used in ExpenseForm
    jest.spyOn(global.Date, 'now').mockImplementation(() =>
        new Date('2019-05-14T11:01:58.135Z').valueOf()
    )
})


test('Render Expense Form', () => {
    const renderer = new ShallowRenderer()
    renderer.render(<ExpenseModal expenses={expenses} openModal={true} />)
    expect(renderer.getRenderOutput()).toMatchSnapshot()
})

test('Render Edit Form', () => {
    const renderer = new ShallowRenderer()
    renderer.render(<ExpenseModal
        openModal={true}
        expenses={expenses}
        expense={
            {
                id: '004',
                description: 'Phone bill',
                note: 'Phone',
                amount: 50,
                createdAt: 1
            }
        }
    />)
    expect(renderer.getRenderOutput()).toMatchSnapshot()
})

test('Check if text update', () => {
    const value = 'My test name'
    const wrapper = shallow(<ExpenseModal openModal={true} />)

    wrapper.find('#item-name').simulate('change', {
        target: { value: value }
    })

    //wrapper.props() returns all the values under node root
    //items are the extract of all the fileds inside Modal
    const props = wrapper.props()
    const items = props.children.props.children

    //items[0] is the first TextFiels, locate the name value
    const item = items[0]
    expect(item.props.value).toBe(value)
    expect(wrapper).toMatchSnapshot()

})

test('Check if date update', () => {
    const value = 'Mon Jul 20 2020 12:34:26 GMT+0300 (Eastern European Summer Time)'
    const wrapper = shallow(<ExpenseModal openModal={true} />)

    wrapper.find('#item-date').simulate('change', value)

    const props = wrapper.props()
    const items = props.children.props.children

    //items[2] is the date
    const item = items[2].props.children.props.value
    expect(item).toBe(value)
    expect(wrapper).toMatchSnapshot()

})
