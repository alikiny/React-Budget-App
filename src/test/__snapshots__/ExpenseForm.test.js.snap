// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check if date update 1`] = `
<ForwardRef(Modal)
  open={true}
>
  <div
    className="makeStyles-paper-1"
  >
    <WithStyles(ForwardRef(TextField))
      id="item-name"
      label="Name"
      onChange={[Function]}
      value=""
      variant="outlined"
    />
    <WithStyles(ForwardRef(TextField))
      id="item-amount"
      label="Amount"
      onChange={[Function]}
      value=""
      variant="outlined"
    />
    <MuiPickersUtilsProvider
      utils={[Function]}
    >
      <PickerWithState
        KeyboardButtonProps={
          Object {
            "aria-label": "set date",
          }
        }
        allowKeyboardControl={true}
        disableToolbar={true}
        format="MM/dd/yyyy"
        id="item-date"
        inputVariant="outlined"
        invalidDateMessage="Invalid Date Format"
        label="Date"
        margin="normal"
        maxDate={2100-01-01T00:00:00.000Z}
        maxDateMessage="Date should not be after maximal date"
        minDate={1900-01-01T00:00:00.000Z}
        minDateMessage="Date should not be before minimal date"
        onChange={[Function]}
        openTo="date"
        value="Mon Jul 20 2020 12:34:26 GMT+0300 (Eastern European Summer Time)"
        views={
          Array [
            "year",
            "date",
          ]
        }
      />
    </MuiPickersUtilsProvider>
    <WithStyles(ForwardRef(TextField))
      id="item-note"
      label="Note"
      onChange={[Function]}
      value=""
      variant="outlined"
    />
    <div>
      <WithStyles(ForwardRef(IconButton))
        id="save-btn"
        onClick={[Function]}
      >
        <AddCircleIcon
          className="makeStyles-addIcon-2"
        />
      </WithStyles(ForwardRef(IconButton))>
      <WithStyles(ForwardRef(IconButton))
        id="close-btn"
        onClick={[Function]}
      >
        <CancelIcon
          className="makeStyles-addIcon-2"
        />
      </WithStyles(ForwardRef(IconButton))>
    </div>
  </div>
</ForwardRef(Modal)>
`;

exports[`Check if text update 1`] = `
<ForwardRef(Modal)
  open={true}
>
  <div
    className="makeStyles-paper-1"
  >
    <WithStyles(ForwardRef(TextField))
      id="item-name"
      label="Name"
      onChange={[Function]}
      value="My test name"
      variant="outlined"
    />
    <WithStyles(ForwardRef(TextField))
      id="item-amount"
      label="Amount"
      onChange={[Function]}
      value=""
      variant="outlined"
    />
    <MuiPickersUtilsProvider
      utils={[Function]}
    >
      <PickerWithState
        KeyboardButtonProps={
          Object {
            "aria-label": "set date",
          }
        }
        allowKeyboardControl={true}
        disableToolbar={true}
        format="MM/dd/yyyy"
        id="item-date"
        inputVariant="outlined"
        invalidDateMessage="Invalid Date Format"
        label="Date"
        margin="normal"
        maxDate={2100-01-01T00:00:00.000Z}
        maxDateMessage="Date should not be after maximal date"
        minDate={1900-01-01T00:00:00.000Z}
        minDateMessage="Date should not be before minimal date"
        onChange={[Function]}
        openTo="date"
        value={2019-05-14T11:01:58.135Z}
        views={
          Array [
            "year",
            "date",
          ]
        }
      />
    </MuiPickersUtilsProvider>
    <WithStyles(ForwardRef(TextField))
      id="item-note"
      label="Note"
      onChange={[Function]}
      value=""
      variant="outlined"
    />
    <div>
      <WithStyles(ForwardRef(IconButton))
        id="save-btn"
        onClick={[Function]}
      >
        <AddCircleIcon
          className="makeStyles-addIcon-2"
        />
      </WithStyles(ForwardRef(IconButton))>
      <WithStyles(ForwardRef(IconButton))
        id="close-btn"
        onClick={[Function]}
      >
        <CancelIcon
          className="makeStyles-addIcon-2"
        />
      </WithStyles(ForwardRef(IconButton))>
    </div>
  </div>
</ForwardRef(Modal)>
`;

exports[`Render Edit Form 1`] = `
<ForwardRef(Modal)
  open={true}
>
  <div
    className="makeStyles-paper-1"
  >
    <WithStyles(ForwardRef(TextField))
      id="item-name"
      label="Name"
      onChange={[Function]}
      variant="outlined"
    />
    <WithStyles(ForwardRef(TextField))
      id="item-amount"
      label="Amount"
      onChange={[Function]}
      value={50}
      variant="outlined"
    />
    <MuiPickersUtilsProvider
      utils={[Function]}
    >
      <PickerWithState
        KeyboardButtonProps={
          Object {
            "aria-label": "set date",
          }
        }
        allowKeyboardControl={true}
        disableToolbar={true}
        format="MM/dd/yyyy"
        id="item-date"
        inputVariant="outlined"
        invalidDateMessage="Invalid Date Format"
        label="Date"
        margin="normal"
        maxDate={2100-01-01T00:00:00.000Z}
        maxDateMessage="Date should not be after maximal date"
        minDate={1900-01-01T00:00:00.000Z}
        minDateMessage="Date should not be before minimal date"
        onChange={[Function]}
        openTo="date"
        views={
          Array [
            "year",
            "date",
          ]
        }
      />
    </MuiPickersUtilsProvider>
    <WithStyles(ForwardRef(TextField))
      id="item-note"
      label="Note"
      onChange={[Function]}
      value="Phone"
      variant="outlined"
    />
    <div>
      <WithStyles(ForwardRef(IconButton))
        id="save-btn"
        onClick={[Function]}
      >
        <Memo(AddCircleIcon)
          className="makeStyles-addIcon-2"
        />
      </WithStyles(ForwardRef(IconButton))>
      <WithStyles(ForwardRef(IconButton))
        id="close-btn"
        onClick={[Function]}
      >
        <Memo(CancelIcon)
          className="makeStyles-addIcon-2"
        />
      </WithStyles(ForwardRef(IconButton))>
    </div>
  </div>
</ForwardRef(Modal)>
`;

exports[`Render Expense Form 1`] = `
<ForwardRef(Modal)
  open={true}
>
  <div
    className="makeStyles-paper-1"
  >
    <WithStyles(ForwardRef(TextField))
      id="item-name"
      label="Name"
      onChange={[Function]}
      value=""
      variant="outlined"
    />
    <WithStyles(ForwardRef(TextField))
      id="item-amount"
      label="Amount"
      onChange={[Function]}
      value=""
      variant="outlined"
    />
    <MuiPickersUtilsProvider
      utils={[Function]}
    >
      <PickerWithState
        KeyboardButtonProps={
          Object {
            "aria-label": "set date",
          }
        }
        allowKeyboardControl={true}
        disableToolbar={true}
        format="MM/dd/yyyy"
        id="item-date"
        inputVariant="outlined"
        invalidDateMessage="Invalid Date Format"
        label="Date"
        margin="normal"
        maxDate={2100-01-01T00:00:00.000Z}
        maxDateMessage="Date should not be after maximal date"
        minDate={1900-01-01T00:00:00.000Z}
        minDateMessage="Date should not be before minimal date"
        onChange={[Function]}
        openTo="date"
        value={2019-05-14T11:01:58.135Z}
        views={
          Array [
            "year",
            "date",
          ]
        }
      />
    </MuiPickersUtilsProvider>
    <WithStyles(ForwardRef(TextField))
      id="item-note"
      label="Note"
      onChange={[Function]}
      value=""
      variant="outlined"
    />
    <div>
      <WithStyles(ForwardRef(IconButton))
        id="save-btn"
        onClick={[Function]}
      >
        <Memo(AddCircleIcon)
          className="makeStyles-addIcon-2"
        />
      </WithStyles(ForwardRef(IconButton))>
      <WithStyles(ForwardRef(IconButton))
        id="close-btn"
        onClick={[Function]}
      >
        <Memo(CancelIcon)
          className="makeStyles-addIcon-2"
        />
      </WithStyles(ForwardRef(IconButton))>
    </div>
  </div>
</ForwardRef(Modal)>
`;
